igor-web/src/main/groovy/com/netflix/spinnaker/igor/gitlabci/GitlabCiBuildMonitor.java

```diff
@@ -44,64 +51,88 @@
-                            sendEventForPipeline(project, pipeline, gitlabCiService.getAddress(), branchedRepoSlug, master);
-                        }
-                    }
-                },
-                throwable -> log.error(String.format("Error: %s (master: %s)", throwable, kv("master", master)), throwable));
-            if (updatedBuilds.get() > 0) {
-                log.info("Found {} new builds (master: {})", updatedBuilds.get(), kv("master", master));
+        final List<Project> projects = gitlabCiService.getProjects();
+        log.info("Took {} ms to retrieve {} repositories (master: {})", System.currentTimeMillis() - startTime, projects.size(), kv("master", master));
+
+        List<BuildDelta> delta = new ArrayList<>();
+        projects.parallelStream().forEach(project -> {
+            List<Pipeline> pipelines = filterOldPipelines(gitlabCiService.getPipelines(project, MAX_NUMBER_OF_PIPELINES));
+            for (Pipeline pipeline : pipelines) {
+                String branchedRepoSlug = GitlabCiPipelineUtis.getBranchedPipelineSlug(project, pipeline);
+
+                boolean isPipelineRunning = GitlabCiResultConverter.running(pipeline.getStatus());
+                int cachedBuildId = buildCache.getLastBuild(master, branchedRepoSlug, isPipelineRunning);
+                // In case of Gitlab CI the pipeline ids are increasing so we can use it for ordering
+                if (pipeline.getId() > cachedBuildId) {
+                    updatedBuilds.incrementAndGet();
+                    delta.add(new BuildDelta(branchedRepoSlug, project, pipeline, isPipelineRunning));
+                }
             }
+        });
 
-            log.info("Last poll took {} ms (master: {})", System.currentTimeMillis() - startTime, kv("master", master));
-        } catch (Exception e) {
-            log.error("Failed to obtain the list of projects", e);
+        if (delta.size() > 0) {
+            log.info("Found {} new builds (master: {})", updatedBuilds.get(), kv("master", master));
         }
+
+        return new BuildPollingDelta(delta, master, startTime);
+    }
+
+    @Override
+    protected void commitDelta(BuildPollingDelta delta) {
+        int ttl = buildCacheJobTTLSeconds();
+        final GitlabCiService gitlabCiService = (GitlabCiService) buildMasters.getMap().get(delta.master);
+
+        delta.items.parallelStream().forEach(item -> {
+            log.info("Build update [{}:{}:{}] [status:{}] [running:{}]", kv("master", delta.master), item.branchedRepoSlug, item.pipeline.getId(), item.pipeline.getStatus(), item.pipelineRunning);
+            buildCache.setLastBuild(delta.master, item.branchedRepoSlug, item.pipeline.getId(), item.pipelineRunning, ttl);
+            buildCache.setLastBuild(delta.master, item.project.getPathWithNamespace(), item.pipeline.getId(), item.pipelineRunning, ttl);
```

Comment:
why the double `setLastBuild` call?

Reply:
From what I understand, GitlabCI tracks builds at both the project and branch level, so we need to track two different builds.

Reply:
👍